File: elixir/world.ex
defmodule World do
  defstruct [
    :width,
    :height,
    tick: 0,
    cells: %{}
  ]

  defmodule LocationOccupied do
    defexception [:x, :y]
    def message(e), do:
      "LocationOccupied(#{e.x}-#{e.y})"
  end

  @directions [
    {-1, 1},  {0, 1},  {1, 1}, # above
    {-1, 0},           {1, 0}, # sides
    {-1, -1}, {0, -1}, {1, -1} # below
  ]

  def new(width: width, height: height) do
    world = %World{
      width: width,
      height: height
    }

    world
    |> populate_cells()
    |> prepopulate_neighbours()
  end

  def tick(world) do
    cells =
      for {key, cell} <- world.cells, into: %{} do
        alive_neighbours = Cell.alive_neighbours(cell, world)

        cond do
          not cell.alive and alive_neighbours == 3 ->
            {key, %{cell | alive: true}}

          alive_neighbours < 2 or alive_neighbours > 3 ->
            {key, %{cell | alive: false}}

          true ->
            {key, cell}
        end
      end

    %{world | tick: world.tick + 1, cells: cells}
  end

  def render(world) do
    # The following is the fastest
    for y <- 0..(world.height - 1), reduce: "" do
      rendering ->
        for x <- 0..(world.width - 1), reduce: rendering do
          rendering ->
            cell = cell_at(world, x, y)
            if cell do
              rendering <> Cell.to_char(cell)
            else
              rendering
            end
        end <> "\n"
    end

    # The following is slower
    # Enum.map(0..(world.height - 1), fn y ->
    #   Enum.map(0..(world.width - 1), fn x ->
    #     cell = cell_at(world, x, y)
    #     if cell do
    #       Cell.to_char(cell)
    #     end
    #   end) ++ ["\n"]
    # end)
    # |> Enum.join()

    # The following is slower
    # Enum.map_join(0..(world.height - 1), fn y ->
    #   Enum.map_join(0..(world.width - 1), fn x ->
    #     cell = cell_at(world, x, y)
    #     if cell do
    #       Cell.to_char(cell)
    #     end
    #   end) <> "\n"
    # end)
  end

  defp make_key(x, y) do
    # The following is slower
    # "#{x}-#{y}"

    # The following is slower
    # "#{x}" <> "-" <> "#{y}"

    # The following is the fastest
    Enum.join([x, y], "-")
  end

  defp cell_at(world, x, y) do
    key = make_key(x, y)
    Map.get(world.cells, key)
  end

  defp populate_cells(world) do
    for y <- 0..(world.height - 1), reduce: world do
      world ->
        for x <- 0..(world.width - 1), reduce: world do
          world ->
            alive = :rand.uniform() <= 0.2
            {world, _success} = add_cell(world, x, y, alive)
            world
        end
    end
  end

  defp add_cell(world, x, y, alive) do
    existing = cell_at(world, x, y)
    if existing do
      raise LocationOccupied, x: x, y: y
    end

    key = make_key(x, y)
    cell = Cell.new(x, y, alive)
    world = put_in(world.cells[key], cell)
    {world, true}
  end

  defp prepopulate_neighbours(world) do
    for {_key, cell} <- world.cells, reduce: world do
      world ->
        neighbours =
          for {rel_x, rel_y} <- @directions do
            make_key(cell.x + rel_x, cell.y + rel_y)
          end

        put_in(world.cells[make_key(cell.x, cell.y)].neighbours, neighbours)
    end
  end
end

File: erlang/world.erl
-module(world).
-include("cell.hrl").

-export([new/2, dotick/1, render/1, cells/1, tick/1]).
-export_type([world/0]).

-record(world, {
  width :: integer(),
  height :: integer(),
  tick = 0 :: integer(),
  cells = #{} :: #{string() => cell()}
}).
-type world() :: #world{}.

-define(DIRECTIONS, [
  {-1, 1},  {0, 1},  {1, 1}, % above
  {-1, 0},           {1, 0}, % sides
  {-1, -1}, {0, -1}, {1, -1} % below
]).

-spec new(integer(), integer()) -> world().
new(Width, Height) ->
  World = #world{
    width = Width,
    height = Height
  },

  Populated = populate_cells(World),
  prepopulate_neighbours(Populated).

-spec dotick(world()) -> world().
dotick(World) ->
  Cells = maps:map(
    fun(_Key, Cell) ->
      AliveNeighbours = cell:alive_neighbours(Cell, World),
      if
        not Cell#cell.alive andalso AliveNeighbours =:= 3 ->
          Cell#cell{alive = true};
        AliveNeighbours < 2 orelse AliveNeighbours > 3 ->
          Cell#cell{alive = false};
        true ->
          Cell
      end
    end,
    World#world.cells
  ),

  World#world{tick = World#world.tick + 1, cells = Cells}.

-spec render(world()) -> string().
render(World) ->
  %% The following is the fastest
  lists:flatten([
    begin
      Line = [
        begin
          Cell = cell_at(World, X, Y),
          cell:to_char(Cell)
        end
        || X <- lists:seq(0, World#world.width - 1)
      ],
      Line ++ "\n"
    end
    || Y <- lists:seq(0, World#world.height - 1)
  ]).

  %% The following is slower
  %% lists:foldl(
  %%   fun(Y, Rendering) ->
  %%     Line = lists:foldl(
  %%       fun(X, LineAcc) ->
  %%         Cell = cell_at(World, X, Y),
  %%         LineAcc ++ [cell:to_char(Cell)]
  %%       end,
  %%       [],
  %%       lists:seq(0, World#world.width - 1)
  %%     ),
  %%     Rendering ++ Line ++ "\n"
  %%   end,
  %%   [],
  %%   lists:seq(0, World#world.height - 1)
  %% ).

-spec cells(world()) -> #{string() => cell()}.
cells(#world{cells = Cells}) -> Cells.

-spec tick(world()) -> integer().
tick(#world{tick = Tick}) -> Tick.

-spec make_key(integer(), integer()) -> string().
make_key(X, Y) ->
  %% The following is the fastest
  integer_to_list(X) ++ "-" ++ integer_to_list(Y).

  %% The following is slower
  %% lists:concat([X, "-", Y]).

  %% The following is slower
  %% lists:flatten(io_lib:format("~B-~B", [X, Y])).

-spec cell_at(world(), integer(), integer()) -> cell() | undefined.
cell_at(World, X, Y) ->
  Key = make_key(X, Y),
  maps:get(Key, World#world.cells, undefined).

-spec populate_cells(world()) -> world().
populate_cells(World) ->
  lists:foldl(
    fun(Y, YWorld) ->
      lists:foldl(
        fun(X, XWorld) ->
          Alive = (rand:uniform() =< 0.2),
          {NewWorld, _Success} = add_cell(XWorld, X, Y, Alive),
          NewWorld
        end,
        YWorld,
        lists:seq(0, World#world.width - 1)
      )
    end,
    World,
    lists:seq(0, World#world.height - 1)
  ).

-spec add_cell(world(), integer(), integer(), boolean()) -> {world(), boolean()}.
add_cell(World, X, Y, Alive) ->
  Existing = cell_at(World, X, Y),
  case Existing of
    undefined -> ok;
    _ -> erlang:error({location_occupied, X, Y})
  end,

  Key = make_key(X, Y),
  Cell = cell:new(X, Y, Alive),
  NewWorld = World#world{cells = maps:put(Key, Cell, World#world.cells)},
  {NewWorld, true}.

-spec prepopulate_neighbours(world()) -> world().
prepopulate_neighbours(World) ->
  Cells = maps:map(
    fun(_Key, Cell) ->
      Neighbours = [
        Key
        || {RelX, RelY} <- ?DIRECTIONS,
           Nx <- [Cell#cell.x + RelX],
           Ny <- [Cell#cell.y + RelY],
           not (Nx < 0 orelse Ny < 0),
           not (Nx >= World#world.width orelse Ny >= World#world.height),
           Key <- [make_key(Nx, Ny)],
           maps:is_key(Key, World#world.cells)
      ],
      Cell#cell{neighbours = Neighbours}
    end,
    World#world.cells
  ),

  World#world{cells = Cells}.
